File: data-access-layer/modules/identity/svc.ts
import { pool } from "../../db/db.ts";
import { type IIdentityRepo, PgIdentityRepo } from "./repo.ts";
import type {
  ParentConditionKeys,
  ChildConditionKeys,
} from "../../shared/types/identity.ts";
import type { ParentChild } from "../../shared/types/identity.ts";
import { redisClient } from "../../db/redis-client.ts";
import type { AsyncResponseType } from "../../shared/types/response.ts";

export interface IIdentitySvc {
  doesAccountExist(accountId: string): AsyncResponseType<boolean>;
  getParentConditionKeys(
    userId: string,
  ): AsyncResponseType<ParentConditionKeys>;
  getChildConditionKeys(childId: string): AsyncResponseType<ChildConditionKeys>;
  getAllParentChildren(parentId: string): AsyncResponseType<ParentChild[]>;
}

class IdentitySvc implements IIdentitySvc {
  private repo: IIdentityRepo;
  constructor(repo?: IIdentityRepo) {
    this.repo = repo ?? new PgIdentityRepo(pool, redisClient);
  }

  async doesAccountExist(accountId: string): AsyncResponseType<boolean> {
    return await this.repo.doesAccountExist(accountId);
  }

  async getParentConditionKeys(
    userId: string,
  ): AsyncResponseType<ParentConditionKeys> {
    return await this.repo.getParentConditionKeys(userId);
  }

  async getAllParentChildren(
    parentId: string,
  ): AsyncResponseType<ParentChild[]> {
    return await this.repo.getAllParentChildren(parentId);
  }

  async getChildConditionKeys(
    childId: string,
  ): AsyncResponseType<ChildConditionKeys> {
    return await this.repo.getChildConditionKeys(childId);
  }
}

const identitySvc = new IdentitySvc();
export default identitySvc;

File: apps/backend-python/src/modules/identity/svc.py
from abc import ABC, abstractmethod
from typing import List

from src.modules.identity.repo import IIdentityRepo
from src.shared.types.modules.identity.model import (
    ChildConditionKeys,
    ParentChild,
    ParentConditionKeys,
    PropertyUser,
)
from src.shared.types.response import HTTPErrorResponse


class IIdentitySvc(ABC):
    @abstractmethod
    async def get_parent_condition_keys(
        self,
        user_id: str,
    ) -> HTTPErrorResponse[ParentConditionKeys]:
        pass

    @abstractmethod
    async def get_property_user(
        self,
        property_id: str,
        user_id: str,
    ) -> HTTPErrorResponse[PropertyUser]:
        pass

    @abstractmethod
    async def get_all_parent_children(
        self,
        parent_id: str,
    ) -> HTTPErrorResponse[List[ParentChild]]:
        pass

    @abstractmethod
    async def get_child_condition_keys(
        self,
        child_id: str,
    ) -> HTTPErrorResponse[ChildConditionKeys]:
        pass


class IdentitySvc(IIdentitySvc):
    def __init__(self, repo: IIdentityRepo) -> None:
        self.repo = repo

    async def get_parent_condition_keys(
        self,
        user_id: str,
    ) -> HTTPErrorResponse[ParentConditionKeys]:
        return await self.repo.get_parent_condition_keys(user_id)

    async def get_property_user(
        self,
        property_id: str,
        user_id: str,
    ) -> HTTPErrorResponse[PropertyUser]:
        return await self.repo.get_property_user(
            property_id=property_id, user_id=user_id
        )

    async def get_all_parent_children(
        self,
        parent_id: str,
    ) -> HTTPErrorResponse[List[ParentChild]]:
        return await self.repo.get_all_parent_children(parent_id=parent_id)

    async def get_child_condition_keys(
        self,
        child_id: str,
    ) -> HTTPErrorResponse[ChildConditionKeys]:
        return await self.repo.get_child_condition_keys(child_id=child_id)
